### PR TITLE
Finished DDNS project

### PR DESCRIPTION
**Pull request by stringie**

### CODE REVIEW

File: contracts/libs/DomainLib.sol [hunk 1]
```diff
@@ -0,0 +1,37 @@
+pragma solidity ^0.4.23;
+
+library DomainLib {
```
Comment (ArnaudovSt) [inline contracts/libs/DomainLib.sol:3]: Make functions internal.

File: contracts/DDNS.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        require(_ip != 0);
+
+        DomainLib.DomainInfo storage domain = domains[_domain];
+
+        uint price;
+        bool isExtending = false;
+
+        if (domain.exists() && !domain.hasExpired()) {
+            require(msg.sender == domain.getOwner());
+            price = REGISTRATION_FEE;
```
Comment (ArnaudovSt) [inline contracts/DDNS.sol:42]: So if we extend the expiry date of a short-named domain we will pay only the regular fee, but not the bigger one we paid, when we registered the short domain at first?

File: contracts/DDNS.sol [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    function getIP(bytes _domain) public view returns (bytes4) {
+        require(domains[_domain].exists());
+
+        return domains[_domain].ip;
+    }
+    
+    function getPrice(bytes _domain) public view returns (uint) {
+        require(_domain.length > MIN_DOMAIN_NAME_LENGTH);
+        return REGISTRATION_FEE + _domain.length * (100 finney);
```
Comment (ArnaudovSt) [inline contracts/DDNS.sol]: Was not the task saying that the shorter domain should have bigger price?

File: contracts/common/SafeMath.sol [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  /**
+  * @dev Adds two numbers, throws on overflow.
+  */
+  function add(uint256 a, uint256 b) internal pure returns (uint256 c) {
+    c = a + b;
+    assert(c >= a);
+    return c;
+  }
+
+  function divmul(uint256 a, uint256 b, uint256 c) internal pure returns (uint256) {
```
Comment (ArnaudovSt) [inline contracts/common/SafeMath.sol]: You may wish to remove the redundant functions when we deploy in production, even from the libraries.

**Review by ArnaudovSt — CHANGES_REQUESTED** [review]

File: contracts/DDNS.sol (hunk 2 above)
Comment (stringie) [reply contracts/DDNS.sol:42]: Yes, that is how I understood the extension procedure. If you just want to keep your domain you pay one more ETH every year. Only the first time it cost the actual amount for the domain. Is my assumption wrong? Why should we pay the same price again when we first bought the domain just to keep the domain? Maybe my business logic is flawed :D

File: contracts/DDNS.sol (hunk 3 above)
Comment (stringie) [reply contracts/DDNS.sol]: You're right, I misread it. Will fix.

File: contracts/common/SafeMath.sol (hunk 4 above)
Comment (stringie) [reply contracts/common/SafeMath.sol]: Forgot to do so when copied from previous project.

**Review by ArnaudovSt — APPROVED** [review]

**Pull request merged**